haystack/utils/device.py

```diff
@@ -1,34 +1,471 @@
+    a single device or a device map.
+    """
+
+    _single_device: Optional[Device] = field(default=None)
+    _multiple_devices: Optional[DeviceMap] = field(default=None)
+
+    @classmethod
+    def from_str(cls, device_str: str) -> "ComponentDevice":
+        """
+        Create a component device representation from a device string.
+        The device string can only represent a single device.
+
+        :param device_str:
+            The device string.
+        :returns:
+            The component device representation.
+        """
+        device = Device.from_str(device_str)
+        return cls.from_single(device)
+
+    @classmethod
+    def from_single(cls, device: Device) -> "ComponentDevice":
+        """
+        Create a component device representation from a single device.
+        Disks cannot be used as single devices.
+
+        :param device:
+            The device.
+        :returns:
+            The component device representation.
+        """
+        if device.type == DeviceType.DISK:
+            raise ValueError("The disk device can only be used as a part of device maps")
+
+        return cls(_single_device=device)
+
+    @classmethod
+    def from_multiple(cls, device_map: DeviceMap) -> "ComponentDevice":
+        """
+        Create a component device representation from a device map.
+
+        :param device_map:
+            The device map.
+        :returns:
+            The component device representation.
+        """
+        return cls(_multiple_devices=device_map)
+
+    def to_torch(self) -> "torch.device":
+        """
+        Convert the component device representation to PyTorch format.
+        Device maps are not supported.
+
+        :returns:
+            The PyTorch device representation.
+        """
+        if self._single_device is None:
+            raise ValueError("Only single devices can be converted to PyTorch format")
+
+        torch_import.check()
+        assert self._single_device is not None
+        return torch.device(str(self._single_device))
+
+    def to_torch_str(self) -> str:
+        """
+        Convert the component device representation to PyTorch string format.
+        Device maps are not supported.
+
+        :returns:
+            The PyTorch device string representation.
+        """
+        if self._single_device is None:
+            raise ValueError("Only single devices can be converted to PyTorch format")
+
+        assert self._single_device is not None
+        return str(self._single_device)
+
+    def to_spacy(self) -> int:
+        """
+        Convert the component device representation to spaCy format.
+        Device maps are not supported.
+
+        :returns:
+            The spaCy device representation.
+        """
+        if self._single_device is None:
+            raise ValueError("Only single devices can be converted to spaCy format")
+
+        assert self._single_device is not None
```

Comment:
Is this assert needed since we are checking for `None` above in the if statement?

Reply:
It's just there to pacify `mypy` at the moment (it complains about a potential `Optional` otherwise).